draft/ZEP0010.md

```diff
@@ -0,0 +1,440 @@
+---
+layout: default
+title: ZEP010
+description: This ZEP proposes a new generic extensions field.
+parent: draft ZEPs
+nav_order: 10
+---
+
+# ZEP 10 — Zarr Generic Extensions
+
+Authors:
+
+- [Norman Rzepka](https://github.com/normanrz), scalable minds
+- [Josh Moore](https://github.com/joshmoore), German BioImaging
+
+Status: Draft
+
+Type: Specification
+
+Created: 2025-05-12
+
+## Abstract
+
+This proposal defines a new generic extension point, ``extensions``, to be
+included in the metadata of Zarr v3 arrays and groups. The ``extensions`` field
+provides a consistent mechanism for attaching additional metadata that does not
+fit into existing extension points defined by the core specification. Extension
+entries within this field follow the naming and structure rules established in
+ZEP0009. This mechanism enables third parties to define and share metadata
+extensions without requiring changes to the core specification or introducing
+new top-level keys.
+
+## Introduction
+
+Zarr specification version 3 currently defines four extension points, each
+associated with a specific (array) metadata field. Additional extension points
+may be added by future ZEPs. Until that time, however, third-parties may want
+to add arbitrary extension objects to either arrays or groups. This proposal
+introduces a generic ``extensions`` field that serves as a container for such a
+list of extensions.
+
+These general purpose extensions are not limited by the scopes of existing
+extension points and require no heavy-weight process to add functionality or alter
+behavior of arrays and groups.
+The intent is to facilitate decentralized and low-friction
+innovation within the Zarr ecosystem by enabling third parties to experiment
+with new features without requiring immediate changes
+to the core specification.
+By tolerating a broader range of experimental extensions, the community can
+explore diverse use cases and patterns. Over time, widely adopted extensions
+may serve as the foundation for future standardization through new ZEPS which
+introduce new extension points or even core features.
+
+## Proposal
+
+To provide for more flexible, immediate, and de-centralized use cases, we
+propose to add a generic extension point ``extensions`` on
+both arrays and groups into which extensions MAY be added.
+
+This field is similar in flexibility to the ``attribues`` field. Conceptually,
+``extensions`` is intended primarily for use by software and automated
+processes, with the potential to influence behavior or processing logic,
+whereas ``attributes`` are generally intended for human interpretation and
+serve as passive metadata or provenance information, though the boundaries are
+not always distinct.
+
+By adding a new field, the specification can assert restrictions that if added
+to ``attributes``. would amount to a breaking change. If present, the
```

Comment:
I don't understand "By adding a new field, the specification can assert restrictions that if added
to ``attributes``. would amount to a breaking change". I think an example could help a lot for people (like me) who've fallen behind in tracking spec/ZEP progress.

Reply:
The contract of "attributes" has been, 'you can put anything here'. The definition of extensions say that they follow a form and their names are registered in https://github.com/zarr-developers/zarr-extensions. Breaking here means that it's certainly fair to assume that there is V3 data today with attributes that would violate these assumptions.